lib/improver/tests/utilities/test_cube_units.py

```diff
@@ -0,0 +1,539 @@
+        cube = self.cube_non_integer_intervals
+        cube_units.DEFAULT_UNITS[diagnostic]['unit'] = target_units
+        cube_units.DEFAULT_UNITS[diagnostic]['dtype'] = np.int32
+
+        msg = ('Data type of diagnostic "air_temperature" could not be'
+               ' enforced without losing significant precision.')
+        with self.assertRaisesRegex(ValueError, msg):
+            self.plugin([cube], check_precision=True)
+
+    def test_temperature_to_invalid_units(self):
+        """Test that a cube with temperatures in kelvin cannot be converted
+        so incompatible units, e.g. metres."""
+
+        target_units = "m"
+        diagnostic = "air_temperature"
+        cube = self.cube
+        cube_units.DEFAULT_UNITS[diagnostic]['unit'] = target_units
+
+        msg = ('Data type of diagnostic "air_temperature" could not be'
+               ' enforced without losing significant precision.')
+        msg = 'air_temperature units cannot be converted to "m"'
+        with self.assertRaisesRegex(ValueError, msg):
+            self.plugin([cube], check_precision=True)
+
+    def test_unavailable_diagnostic(self):
+        """Test application of the function to a cube for which the default
+        units and data type are not defined, resulting in an exception."""
+
+        cube = self.cube
+        cube.rename('number_of_fish')
+
+        msg = 'Diagnostic number_of_fish not defined in units.py'
+        with self.assertRaisesRegex(KeyError, msg):
+            self.plugin([cube])
+
+    def test_return_changes_as_copy(self):
+        """Test that using the inplace=False keyword arg results in the input
+        cube remaining unchanged, and a new modified cube being returned. In
+        this test the temperature are converted to Celsius to make the change
+        clear."""
+
+        target_units = "celsius"
+        diagnostic = "air_temperature"
+        cube = self.cube.copy()
+        cube_units.DEFAULT_UNITS[diagnostic]['unit'] = target_units
+        cube_units.DEFAULT_UNITS[diagnostic]['dtype'] = np.float32
+        expected = np.full((5, 5), -272.15, dtype=np.float32)
+
+        result, = self.plugin([cube], inplace=False)
+
+        self.assertArrayEqual(cube.data, self.cube.data)
+        self.assertEqual(cube.units, self.cube.units)
+
+        self.assertArrayEqual(result.data, expected)
+        self.assertEqual(result.units, target_units)
+        self.assertEqual(result.data.dtype, np.float32)
+
+    def test_temperature_to_integer_celsius_valid(self):
+        """Test that a cube with temperatures at whole kelvin intervals can
+        be converted to integer Celsius. Precision checking is not employed
+        here as the values change by the relevant conversion and may not end up
+        as integers, making such a check unhelpful. This is not a likely use
+        case unless precision loss is intentional; here the 0.15 Celsius us
+        lost from the resulting temperatures."""
```

Comment:
Yeah, you are right I think. This was effectively adding the ability to clip the data, but we have tools to explicitly do that from numpy etc, so I'll change the behaviour.

Reply:
I've altered this test to convert to celsius as a float from kelvin as integers. This shows changing both aspects simultaneously as an extra test. Now that I've removed the option to avoid checking for precision loss for diagnostics (i.e. it's mandatory) the error you suggest raising is done in that check; this is shown in test_temperature_to_integer_kelvin_invalid above.

Reply:
Thanks.